File: src/main/java/com/lkimilhol/matchingproject/address/domain/Address.java
package com.lkimilhol.matchingproject.address.domain;

import java.util.Objects;

import lombok.*;

import javax.persistence.*;

import com.lkimilhol.matchingproject.member.domain.Member;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Address {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "address_id")
    private Long id;

    @Enumerated(EnumType.STRING)
    private City city;        //시 ex) 서울시, 부산광역시, 광주광역시

    @Embedded
    private District district;    //구 ex) 송파구, 강동

    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    @JoinColumn(name = "member_id")
    private Member member;

    public Address(City city, District district, Member member) {
        this.city = city;
        this.district = district;
        this.member = member;
    }

    public static Address of(City city, District district, Member member) {
        return new Address(city, district, member);
    }

    public void update(City city, District district) {
        this.city = city;
        this.district = district;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Address address = (Address) o;
        return Objects.equals(id, address.id) && city == address.city && Objects.equals(district, address.district) && Objects.equals(member, address.member);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, city, district, member);
    }
}

File: src/main/kotlin/com/lkimilhol/matchingproject/address/domain/Address.kt
package com.lkimilhol.matchingproject.address.domain

import com.lkimilhol.matchingproject.member.domain.Member
import javax.persistence.*

@Entity
class Address(
    //시 ex) 서울시, 부산광역시, 광주광역시
    @Enumerated(EnumType.STRING)
    var city: City,

    //구 ex) 송파구, 강동
    @Embedded
    var district: District,

    @ManyToOne(fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    @JoinColumn(name = "member_id")
    val member: Member? = null
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "address_id")
    val id: Long? = null

    fun update(city: City, district: District) {
        this.city = city
        this.district = district
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Address

        if (id != other.id) return false
        if (city != other.city) return false
        if (district != other.district) return false
        if (member != other.member) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id?.hashCode() ?: 0
        result = 31 * result + (city?.hashCode() ?: 0)
        result = 31 * result + (district?.hashCode() ?: 0)
        result = 31 * result + (member?.hashCode() ?: 0)
        return result
    }
}